### PR TITLE
Fix warnings about dollar and dlsym 

### PR DESCRIPTION
**Pull request by recp**
We were getting some warnings about dlsym when compiling like warnings in the issues #1447 and #1444

C does not support casts between function pointers and data pointers (yet) because they sizes can be different.... Maybe there are some optimizations such as implicit casts when compiling/linking for dlsym. This can be used until found better solution

There are some solutions here: http://christopherpoole.github.io/dynamically-linking-shared-objects-in-c++/ for avoiding void \* to fn pointer.

There is a comment on this man page http://linux.die.net/man/3/dlsym like this:

``` C
/* Writing: cosine = (double (*)(double)) dlsym(handle, "cos");
       would seem more natural, but the C99 standard leaves
       casting from "void *" to a function pointer undefined.
       The assignment used below is the POSIX.1-2003 (Technical
       Corrigendum 1) workaround; see the Rationale for the
       POSIX specification of dlsym(). */

   *(void **) (&cosine) = dlsym(handle, "cos");
```

I applied the approach/solution for avoiding warnings.
- Fix warnings about dlsym and others
- Fix warnings about dollar

### CODE REVIEW

**Comment by txdv** [comment]
You mean the official C89 standard?

File: .gitignore [hunk 1]
```diff
@@ -64,3 +64,6 @@ ipch
 
 # sphinx generated files
 /docs/build/
+
+*.xcodeproj
+*.xcworkspace
```
Comment (saghul) [inline .gitignore]: Please make this a different commit, it's unrelated change.

File: Makefile.am [hunk 2]
```diff
@@ -81,7 +81,9 @@ else  # WINNT
 
 include_HEADERS += include/uv-unix.h
 AM_CPPFLAGS += -I$(top_srcdir)/src/unix
-libuv_la_CFLAGS += -g --std=gnu89 -pedantic -Wall -Wextra -Wno-unused-parameter
+
+libuv_la_CFLAGS += -g --std=gnu89 -pedantic -Wall -Wextra -Wno-unused-parameter -Wno-dollar-in-identifier-extension
```
Comment (saghul) [inline Makefile.am]: "-Wno-dollar-in-identifier-extension" won't work with compiler other than Clang, AFAIK. Check the comment by Ben in #1447.

**Comment by saghul** [comment]
Please squash the last 2 commits into a single one.

File: Makefile.am [hunk 3]
```diff
@@ -81,7 +81,9 @@ else  # WINNT
 
 include_HEADERS += include/uv-unix.h
 AM_CPPFLAGS += -I$(top_srcdir)/src/unix
-libuv_la_CFLAGS += -g --std=gnu89 -pedantic -Wall -Wextra -Wno-unused-parameter
+
```
Comment (saghul) [inline Makefile.am]: Extraneous newline.

File: Makefile.am (hunk 2 above)
Comment (saghul) [inline Makefile.am]: The flag would also need to be set on GYP builds. Unfortunately I don't know how to check for the compiler name in a GYP file.

**Comment by recp** [comment]
Hi @txdv actually I didn't mean to C89 only. There are some reasons here about why we are getting the warnings about dlsym (data pointers -> fn-ptrs)

[Why are function pointers and data pointers incompatible in C/C++?](http://stackoverflow.com/questions/12358843/why-are-function-pointers-and-data-pointers-incompatible-in-c-c)

File: .gitignore (hunk 1 above)
Comment (recp) [inline .gitignore]: done

**Comment by recp** [comment]
> Please squash the last 2 commits into a single one.

Squashed!

**Comment by saghul** [comment]
GitHub doesn't show my comment anymore, but the "-Wno-dollar-in-identifier-extension" only works with Clang, not GCC. You'll need to detect if the compiler supports it, see Ben's comment in #1447.

Also, that fix would need to be applied for GYP builds.

**Comment by recp** [comment]
> GitHub doesn't show my comment anymore, but the "-Wno-dollar-in-identifier-extension" only works with Clang, not GCC. You'll need to detect if the compiler supports it, see Ben's comment in #1447.

Are we getting warnings about dollar on GCC? I've just compiled libuv-master on Fedora I got no warnings. How can I know/monitor warnings on GCC then (Maybe its related Fedora or Linux)?

**Comment by saghul** [comment]
> Are we getting warnings about dollar on GCC? I've just compiled libuv-master on Fedora I got no warnings. How can I know/monitor warnings on GCC then (Maybe its related Fedora or Linux)?

Those symbols only show up when building with dtrace, so just on OSX. Also, looks like recent GCC may support it, but it wasn't in the past: https://github.com/joyent/libuv/compare/633d33a92a40...0ff06b4ac38a

So we'd need to feature-test it or compilation could fail. Unfortunately I no longer have a OSX 10.6 machine with had GCC 4.2.1.

**Comment by saghul** [comment]
Managed to install the old GCC on OSX Mavericks:

```
(git|pr/1468)saghul@echo.local:~/src/libuv
$ make
  GEN      include/uv-dtrace.h
/Applications/Xcode.app/Contents/Developer/usr/bin/make  all-am
  CC       src/libuv_la-fs-poll.lo
cc1: error: unrecognized command line option "-Wno-dollar-in-identifier-extension"
make[1]: *** [src/libuv_la-fs-poll.lo] Error 1
make: *** [all] Error 2
(git|pr/1468)saghul@echo.local:~/src/libuv
```

**Comment by recp** [comment]
> So we'd need to feature-test it or compilation could fail. Unfortunately I no longer have a OSX 10.6 machine with had GCC 4.2.1.

@saghul can we check the flag in configure.ac like Ben's commit f75bb7c3379d4398de8ed588a110aa7ab8708a91 :

configure.ac

```
CC_CHECK_CFLAGS_APPEND([-Wno-dollar-in-identifier-extension])
```

Ben was created a m4 file for check flags m4/libuv-check-flags.m4 but the commit was reverted by the commit 1d42b3ce7c184696960a96f129ac586749b76a84 :/

Maybe we can re-create this file for checking the flag for avoiding dollar warnings?

/cc @bnoordhuis @indutny

**Comment by saghul** [comment]
That's exactly what we need. That's why I mentioned the note Ben added on
the other issue.

Reminder: gyp :-)

**Comment by saghul** [comment]
I forgot to mention, that while Ben's approach is what we need, we don't want to enable all warnings at the moment, since we need to analyse them on a case by case basis. Please just enable the relevant flag in this case. Thanks!

**Comment by recp** [comment]
I thought so according to Fedor's comments on revert commit: 1d42b3ce7c184696960a96f129ac586749b76a84 ;)

**Comment by recp** [comment]
@saghul I added the flag -Wno-dollar-in-identifier-extension to gyp by checking clang. Im not sure that is working correctly or not, I built on my OS X with gyp and it seems ok

Can you review it?

File: uv.gyp [hunk 4]
```diff
@@ -120,13 +120,28 @@
...
           },
         }, { # Not Windows i.e. POSIX
-          'cflags': [
-            '-g',
-            '--std=gnu89',
-            '-pedantic',
-            '-Wall',
-            '-Wextra',
-            '-Wno-unused-parameter',
+		  'conditions': [
```
Comment (saghul) [inline uv.gyp]: Please use 2 spaces for indentation, like the rest of the file.

Comment (recp) [inline uv.gyp]: Its done. The spaces seems different when working in Xcode and out :/ I re-indented spaces in TextMate.

**Comment by saghul** [comment]
Looks good! I'll check it out a bit later.

**Comment by saghul** [comment]
There seems to be something wrong with the syntax in uv.gyp. None of the flags in the edited block are applied now.

**Comment by saghul** [comment]
This seems like the proper fix:

```
diff --git a/uv.gyp b/uv.gyp
index f121898..3213a05 100644
--- a/uv.gyp
+++ b/uv.gyp
@@ -195,6 +195,10 @@
             '_DARWIN_UNLIMITED_SELECT=1',
           ]
         }],
+        [ 'OS!="win" and clang==1', {
+          # Not all versions of GCC support it, so just enable it on clang
+          'cflags': [ '-Wno-dollar-in-identifier-extension' ],
+        }],
         [ 'OS!="mac"', {
           # Enable on all platforms except OS X. The antique gcc/clang that
           # ships with Xcode emits waaaay too many false positives.
```

**Comment by saghul** [comment]
Hum, I don't get it. None of the flags seem to affect OSX, they are just not there :-O The simplified fix I proposed works on Linux, but with or without the last patch, cflags are not put into the generated makefile :-(

**Comment by recp** [comment]
> ```
>    [ 'OS!="win" and clang==1', {
>       # Not all versions of GCC support it, so just enable it on clang
>       'cflags': [ '-Wno-dollar-in-identifier-extension' ],
>     }],
> ```

You mean its wrong? The gyp puzzled me :-S I thought clang doesnt work on win.

**Comment by saghul** [comment]
The symbols are only present on Unix source files, hence the check.

**Comment by bnoordhuis** [comment]
> None of the flags seem to affect OSX, they are just not there

That's GYP's Xcode emulation mode at work.  You enable the flags like this:

```
xcode_settings': {
  'OTHER_CFLAGS': [ ... ]
}
```

EDIT: Warnings should go in WARNING_CFLAGS, linker flags in OTHER_LDFLAGS and everything else in OTHER_CFLAGS.

**Comment by saghul** [comment]
@bnoordhuis Oh! Does that need to be used in even when makefiles or ninja files are generated instead of xcode project files?

**Comment by bnoordhuis** [comment]
@saghul Yes.

**Comment by saghul** [comment]
How fun, uv.gyp has been fucked up for a while then :-S Thanks a lot for the pointer Ben!

**Comment by recp** [comment]
@bnoordhuis @saghul I found the [tutorial](http://www.rioki.org/2013/02/07/experimenting-with-gyp.html) and there are some xcode_settings like Ben's comment.

if I'll replace prev commit like below will it be correct? I think there is no req OS="mac" anymore :)

``` javascript
{
  'target_defaults': {
    'default_configuration': 'Release',
    'configurations': {
      'Debug': {
         'xcode_settings': {
           'WARNING_CFLAGS': [
              '-Wno-dollar-in-identifier-extension'
           ]
        },

       'Release': {
         'xcode_settings': {
           'WARNING_CFLAGS': [
              '-Wno-dollar-in-identifier-extension'
           ]
        }
```

**Comment by saghul** [comment]
@recp the other flags also need to be set (though you can do that in a separate commit).

**Comment by recp** [comment]
@saghul I updated the gyp to use `xcode_settings` 

It seems work. I checked Clang for dollar-sign for avoiding errors/crashes when building with older Xcode's which use GCC.  
The flags which we set on gyp seems on output when compiling with `xcodebuild` but not with `make` ($ ./gyp_uv.py -f make && make -C out). Also I think the commit can be revert 7e3403e8992ec0970fd160641b8fbd804a5b9521 ?

**Comment by saghul** [comment]
@recp thanks! Yes, that commit should go away. I'll test it throroughly tomorrow, can you please squash your commits?

**Comment by recp** [comment]
I'll squash them which have same title or all into one?

**Comment by saghul** [comment]
Not all, just squash those which are intermediate states.

**Comment by recp** [comment]
Got it, I'll do that tomorrow. Thanks

**Comment by recp** [comment]
@saghul squashed! Can you review it?

**Comment by saghul** [comment]
Will do! Thanks!

File: uv.gyp [hunk 5]
```diff
@@ -26,6 +26,32 @@
...
+          ]}
+        ]
+      ],
+      'OTHER_LDFLAGS': [
+      ],
+      'OTHER_CFLAGS': [
+        '-g',
+        '--std=gnu89',
+        '-pedantic',
+        '-D_DARWIN_USE_64_BIT_INODE=1',
```
Comment (saghul) [inline uv.gyp]: These last 2 (-D_DARWIN...) are not needed, as gyp properly understands them from above, please remove them.

File: src/unix/fsevents.c [hunk 6]
```diff
@@ -525,7 +525,7 @@ static int uv__fsevents_global_init(void) {
   err = -ENOENT;
 #define V(handle, symbol)                                                     \
   do {                                                                        \
-    p ## symbol = dlsym((handle), #symbol);                                   \
+    *(void **)(&p ## symbol) = dlsym((handle), #symbol);                                   \
```
Comment (saghul) [inline src/unix/fsevents.c]: style: please align the \

File: src/unix/darwin-proctitle.c [hunk 7]
```diff
@@ -118,14 +118,14 @@ int uv__set_process_title(const char* title) {
   if (launch_services_bundle == NULL)
     goto out;
 
-  pLSGetCurrentApplicationASN =
-      pCFBundleGetFunctionPointerForName(launch_services_bundle,
+  *(void **)(&pLSGetCurrentApplicationASN) =
+       pCFBundleGetFunctionPointerForName(launch_services_bundle,
```
Comment (saghul) [inline src/unix/darwin-proctitle.c]: extra leading space?

**Comment by saghul** [comment]
Great work @recp! Left some last comments, address those, squash, and we are game!

**Comment by recp** [comment]
@saghul it's done :+1:

**Comment by saghul** [comment]
Kewl looks great, will merge later today!

**Comment by saghul** [comment]
All commits landed on v1.x, thanks a lot for working on this @recp, much appreciated!

**Comment by recp** [comment]
Thanks!
